### ISSUE TITLE
[pytorch/jit] all lists get merged during tracing

### ISSUE THREAD

**Issue by zeryx** [pytorch/pytorch#12780]
## 🐛 Bug

If you define a nn.Module that contains a forward func with multiple lists (That you stack together into a tensor at some point) that process works perfectly fine in the conventional workflow. 
However, if you attempt to convert your module into a ScriptModule that will be compiled into a graph representation, every list object defined in a `script_method` function seem to merge together somehow.

## To Reproduce

Create a test bench with the source code below, it should be fully self contained
```python


import torch
from torch import nn
from torch.functional import F

class ACTNN(torch.jit.ScriptModule):

    __constants__ = ['hidden_depth', 'hidden_width']
    # hidden_width = 125
   # hidden_depth = 3
    # output_size = 80
    # input_size = 125
    def __init__(self, input_size, hidden_width, output_size, hidden_depth):
        super(ACTNN, self).__init__()
        self.hidden_depth = hidden_depth
        self.hidden_width = hidden_width
        self.rnn = torch.jit.trace(nn.RNN(input_size, hidden_width,
                                          num_layers=self.hidden_depth, batch_first=True), example_inputs=(torch.rand(1, 1, input_size), torch.rand(self.hidden_depth, 1, hidden_width)))

        self.proc = torch.jit.trace(nn.Linear(hidden_width, output_size), example_inputs=torch.rand(1, 1, hidden_width))


    @torch.jit.script_method
    def forward(self, input: torch.Tensor, hidden: torch.Tensor):

        input = input.view(1, 1, -1)
        hidden = hidden.view(self.hidden_depth, 1, self.hidden_width)
        # Define the list objects
        states = []
        hiddens = []
        halt_probs = []
        n = 0
        rnn_out, hidden = self.rnn(input, hidden)
        state = self.proc(rnn_out)
        hiddens.append(hidden)
        states.append(state)
        halt_probability = F.sigmoid(hiddens[n].sum())
        halt_probs.append(halt_probability)
       # At this point, both the states & halt_probs lists will contain:
       # 1x 1,1,80 wide tensor
       # 1x 3,1,125 wide tensor
       # 1x 1 wide tensor

        print(states)
        print(halt_probs)
        residual = torch.sum(torch.cat(halt_probs))

        states_tensor = torch.stack(states, dim=1)
        hiddens_tensor = torch.stack(hiddens, dim=1)
        halt_subset = halt_probs[0:n-1]
        halt_subset.append(residual)
        halt_prob_tensor = torch.stack(halt_subset).view(-1)
        output = torch.mv(states_tensor, halt_prob_tensor)
        hidden = torch.mv(hiddens_tensor, halt_prob_tensor)
        return output, hidden
```
dimensions of elements in `states`:
```
[ Variable[CPUFloatType]{3,1,125} ],
[ Variable[CPUFloatType]{1,1,80} ],
[ Variable[CPUFloatType]{1} ]]
```

and here is the error message that causes an overall failure of the trace operation:
```
RuntimeError: 
invalid argument 0: Sizes of tensors must match except in dimension 1. Got 3 and 1 in dimension 0 at /pytorch/aten/src/TH/generic/THTensorMoreMath.cpp:1317:
operation failed in interpreter:
    halt_probability = self.halt_noise(halt_probability, training)
    halt_probs.append(halt_probability)
    print(halt_probs)
    residual = torch.sum(torch.cat(halt_probs))
    states_tensor = torch.stack(states, dim=1)
                    ~~~~~~~~~~~ <--- HERE
```

## Expected behavior

for the above code sample, 
`hiddens` should contain:
```
[ Variable[CPUFloatType]{3,1,125} ],
```
`states` should contain:
```
[ Variable[CPUFloatType]{1,1,80} ]
```
`halt_probs` should contain:
```
[ Variable[CPUFloatType]{1} ]]
```
.
## Environment

 - PyTorch Version (e.g., 1.0): 1.0.0.dev20181004
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): `pip install /tmp/torch_nightly-1.0.0.dev20181004-cp35-cp35m-linux_x86_64.whl`
 - Build command you used (if compiling from source): N/A
 - Python version: 3.5.4
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

## Additional context

As I am doing some somewhat advanced pytorch module development, I could understand if my usage here is not supported - however there is currently no error message saying that I can't construct multiple tensors from elements captured in a forward pass.

It's certainly possible that some other part of my algorithm is failing, or maybe I'm using torch.jit.ScriptModule incorrectly? That would be great if that was the case,

I could see this posing issues down the road with other types of dynamic process flow during forward pass loops (like ACT, but also NTMs, etc), which are probably mechanisms we want pytorch 1.0 to support properly.

**Comment by apaszke** [comment]
It doesn't seem like you have anything unreasonable in your code. It's simply likely a bug on our side.

**Comment by phi-go** [comment]
I'm facing a similar bug. Minimal code example:

```python
import torch

from torch.jit import script


@script
def update_tensor(tensor):
    tensor_prime = []
    for i in range(len(tensor)):
        n = tensor[i]
        tensor_prime.append(n)
    nodes_prime = torch.stack(tensor_prime)
    return nodes_prime


def main():
    tensor = torch.FloatTensor(4, 2).random_()

    for i in range(3):
        print("="*20)
        print(tensor.shape)
        print(tensor)
        tensor = update_tensor(tensor)


if __name__ == "__main__":
    main()
```

Output with PYTORCH_JIT=1:

```
====================
torch.Size([4, 2])
tensor([[ 1808538.,  7064277.],
        [  239404., 11649655.],
        [ 1407555., 12730792.],
        [11626145.,  6317278.]])
====================
torch.Size([1, 2])
tensor([[1808538., 7064277.]])
====================
torch.Size([1, 2])
tensor([[1808538., 7064277.]])
```

Output with PYTORCH_JIT=0:

```
====================
torch.Size([4, 2])
tensor([[ 9087322., 11016173.],
        [ 4673285.,  9341248.],
        [ 1091294.,  1047240.],
        [ 6966788., 12850260.]])
====================
torch.Size([4, 2])
tensor([[ 9087322., 11016173.],
        [ 4673285.,  9341248.],
        [ 1091294.,  1047240.],
        [ 6966788., 12850260.]])
====================
torch.Size([4, 2])
tensor([[ 9087322., 11016173.],
        [ 4673285.,  9341248.],
        [ 1091294.,  1047240.],
        [ 6966788., 12850260.]])
```

Seems like stack only takes the first tensor in the list and ignores the others.

Version: `https://download.pytorch.org/whl/nightly/cu92/torch_nightly-1.0.0.dev20181008-cp36-cp36m-linux_x86_64.whl`

**Comment by michaelsuo** [comment]
Taking a look at this now

**Comment by michaelsuo** [comment]
@phi-go your example is caused by a bug in how we do argument promotion. Specifically, `len(tensor)` is returning 1, because `tensor` is being promoted to a list of tensors of size 1. A minimal example to demonstrate the problem:
```
import torch

tensor = torch.IntTensor(4, 2).zero_()
print(len(tensor))  # prints 4

@torch.jit.script
def foo(tensor):
    print(len(tensor)) # this is dispatching to aten::len(Tensor[]), not aten::len(Tensor)

foo(tensor)  # prints 1
```
This is definitely a bug on our side and I'll issue fix. @zeryx, your problem may be different; I'll dig into that as well.

**Comment by phi-go** [comment]
@michaelsuo That explains it, thank you.
Using `range(tensor.size(0))` works as expected.

**Comment by phi-go** [comment]
I was trying to get my code to work with tensor.size but now I'm also experiencing the "list merging" problem. Note that the error only occurs when the `ItsAList` Modules are somehow used, see the comment. Also if the list variable is created in a `@script` function it works as expected, so `ScriptModule`  has to play some role.

Code:
```python
import torch
from torch import nn

from torch.jit import ScriptModule, script_method


class ItsAList(ScriptModule):

    def __init__(self):
        super(ItsAList, self).__init__()

    @script_method
    def forward(self, tensor):
        just_a_list = []
        return just_a_list


class RebuildSeveralTimes(ScriptModule):
    __constants__ = ['module_list', 'times']

    def __init__(self, times):
        super(RebuildSeveralTimes, self).__init__()
        self.times = times
        alist_modules = []
        for i in range(times):
            alist_modules.append(ItsAList())
        self.module_list = nn.ModuleList(alist_modules)

    @script_method
    def forward(self, tensor):
        for _ in self.module_list:  # works as expected with: for _ in range(self.times):
            tensor_list = []
            for i in range(tensor.size(0)):
                n = tensor[i]
                tensor_list.append(n)
            tensor = torch.stack(tensor_list)
        return tensor


for i in range(4):
    tensor = torch.FloatTensor([i])
    model = RebuildSeveralTimes(i)
    new_tensor = model(tensor)
    print(f"i {i}", "="*20)
    print(tensor)
    print(new_tensor)
```

Output:
```
i 0 ====================
tensor([0.])
tensor([0.])
i 1 ====================
tensor([1.])
tensor([1.])
i 2 ====================
tensor([2.])
tensor([2., 2.])
i 3 ====================
tensor([3.])
tensor([3., 3., 3., 3.])
```

Output with PYTHON_JIT=0:
```
i 0 ====================
tensor([0.])
tensor([0.])
i 1 ====================
tensor([1.])
tensor([1.])
i 2 ====================
tensor([2.])
tensor([2.])
i 3 ====================
tensor([3.])
tensor([3.])
```

**Comment by gfjiangly** [comment]
Has this problem been fixed in the pytorch 1.0.1 version?
I encounter this problem(all lists get merged during tracing, like @zeryx's problem) now. But I don't know how to disable optimization.

**Comment by suo** [comment]
This problem should be fixed in 1.0.1; can you provide a small script that we can run to reproduce your issue?
